### ISSUE TITLE
Starting with testing

### ISSUE THREAD

**Issue by to-sta** [activist-org/activist#432]
### Terms

- [X] I have searched [open and closed issues](https://github.com/activist-org/activist/issues?q=is%3Aissue)
- [X] I agree to follow activist's [Code of Conduct](https://github.com/activist-org/activist/blob/main/.github/CODE_OF_CONDUCT.md)

### Issue

# Let's start with testing!

We can use this issue to discuss **testing** in general and spin off **new issues** to work on.

![swiss_cheese_model](https://github.com/activist-org/activist/assets/49518857/86daece7-3894-484f-b341-f5668c12e252.png?raw=true)

Here is an illustration based on the [swiss-cheese-model](https://en.wikipedia.org/wiki/Swiss_cheese_model) of our current **development process** (excluding the testing barrier 😄). The image illustrates how bugs could get into our codebase and what barriers are in the way.

## Topics
1. Splitting development and production.
2. Testing
3. GitHub Workflows

## 1. Splitting development and production
That would involve splitting the `settings.py` file, for example, into `base.py`, `development.py`, and `production.py`, as well as separating the `requirements.txt`. We could use the `env. variable DEBUG` to switch between settings. The development settings should contain the location for the fixtures and a seperate test database. 

## 2. Testing
To start, I would say we should run `coverage.py` to find out where we need tests and then start implementing them. We could work on fixtures asynchronously. For end-to-end testing, we could use [playwright](https://playwright.dev/python/).

## 3. GitHub Workflows
After implementing tests, there are a few additions that can be made to improve the development process. First, we can add test coverage to determine where we need to add new tests. We also have to execute tests in the workflows to find issues that arise during changes. Additionally, I think we should also add a workflow for the build process to find dependency issues.

Since we are have not reached production, i would exclude logging and performance testing in general for now.
Let me know if i have missed something? Or if you have any suggestions. 

@andrewtavis @wkyoshida

**Comment by wkyoshida** [comment]
> To start, I would say we should run `coverage.py` to find out where we need tests and then start implementing them.

Do we already have any? Not sure if I'm just grossly missing them :laughing:, but maybe safe to say that all-around tests needed?

> First, we can add test coverage to determine where we need to add new tests

+1 yep! This is great. What I've often done is have the workflow check if the coverage has gotten worse after the incoming changes. This could just be a warning in the PR or it could fail the check if we'd like to too :eyes::astonished:

**Comment by to-sta** [comment]
Yeah, at least one test is need before we can run coverage 😄, we either add one or wait for #421 to be merged.

> +1 yep! This is great. What I've often done is have the workflow check if the coverage has gotten worse after the incoming changes. This could just be a warning in the PR or it could fail the check if we'd like to too 👀😲

That would be ideal 🤯! I think a warning would be good enough for the start.

**Comment by to-sta** [comment]
Regarding **1. Splitting development and production** requirements.txt, I found [pip-tools](https://github.com/jazzband/pip-tools) as a solution. We seperate the `requirements.txt` file into `requirements.in` and `requirements-dev.in`. 

After that we can use pip-compile to pin the package versions into `requirements.txt` and `requirements-dev.txt` files. 

And finally we need to adjust the pr_ci_backend workflow (after it is merged😄). 

I was also thinking about the dockerfile, i would suggest that we use the requirements-dev.txt in there and create a seperate dockerfile for production later on.

**Comment by andrewtavis** [comment]
Ping @momanyisamuel :)

**Comment by andrewtavis** [comment]
Pinging you both on #427 where we're discussing Storybook that could also be used for frontend testing 🙃

**Comment by to-sta** [comment]
This is the current coverage report made with the PR#496.

```bash
Name                                    Stmts   Miss  Cover
-----------------------------------------------------------
authentication\__init__.py                  0      0   100%
authentication\admin.py                    10      0   100%
authentication\apps.py                      4      0   100%
authentication\migrations\__init__.py       0      0   100%
authentication\models.py                   49      6    88%
authentication\serializers.py              61     61     0%
authentication\tests.py                     1      0   100%
authentication\urls.py                     12     12     0%
authentication\views.py                    28     28     0%
backend\__init__.py                         0      0   100%
backend\asgi.py                             4      4     0%
backend\paginator.py                        5      5     0%
backend\settings.py                        36      0   100%
backend\urls.py                             4      4     0%
backend\wsgi.py                             4      4     0%
content\__init__.py                         0      0   100%
content\admin.py                            7      0   100%
content\apps.py                             4      0   100%
content\factory.py                         41      0   100%
content\migrations\__init__.py              0      0   100%
content\models.py                          47      5    89%
content\serializers.py                     63     63     0%
content\tests.py                           12      8    33%
content\urls.py                            11     11     0%
content\views.py                           24     24     0%
entities\__init__.py                        0      0   100%
entities\admin.py                          15      0   100%
entities\apps.py                            4      0   100%
entities\migrations\__init__.py             0      0   100%
entities\models.py                        100     13    87%
entities\serializers.py                   123    123     0%
entities\tests.py                           1      0   100%
entities\urls.py                           19     19     0%
entities\views.py                          83     83     0%
events\__init__.py                          0      0   100%
events\admin.py                            12      0   100%
events\apps.py                              4      0   100%
events\migrations\__init__.py               0      0   100%
events\models.py                           74      9    88%
events\serializers.py                     102    102     0%
events\tests.py                             1      0   100%
events\urls.py                             16     16     0%
events\views.py                            44     44     0%
manage.py                                  12     12     0%
utils\__init__.py                           0      0   100%
utils\utils.py                             21     21     0%
-----------------------------------------------------------
TOTAL                                    1058    677    36%
```

I think we have to adjust the configuration in the `pyproject.toml` or in `.coveragerc` file to exclude some files (e.g manage.py, wsgi.py and migrations) 

@wkyoshida do you want to create an issue for the coverage workflow? Let me know if I can help you.

**Comment by lpil** [comment]
Hey folks! What's the status here? We're just doing some due diligence on this project and the lack of any unit testing is a red light for us. Would be good to understand how quality and correctness is going to be verified seeing as this project is too large for manual testing.

**Comment by andrewtavis** [comment]
Hey @lpil! Thanks for reaching out on this. There's been a good amount of discussion about this in our dev syncs recently as well as in the community in general. @lucianmurmurache is going to be picking up cypress/ace-core testing for the frontend (at least the system and the tests will have issues made for them), and backend tests are progressing and will get into full swing once we start putting things together in the coming weeks. Any suggestions or further comments would be welcome 😊

**Comment by andrewtavis** [comment]
Post #929 backend tests are in PRs :) @aasimsyed, once we're done with the initial page tests for the frontend and have the fixtures being generated properly, we can then maybe make a few issues for some of the MVP pages. With those being closed we could then close this one 😇

**Comment by andrewtavis** [comment]
Assigning @cquinn540 to give credit for all the work that's been done for e2e testing :) I think for the initial task we're getting close and from there we need to iterate. Maybe a coverage report in the backend ci workflow so we can keep an eye on it?

**Comment by andrewtavis** [comment]
I just made #1121 to add in a coverage report to PRs that change files in the backend. I think that with that we have a pretty solid base of backend, frontend, e2e and i18n tests and can say that we have "started with testing". I'm going to close this :) Thanks all for the great work here and the dedication to building this out in the best way possible! 😊
